OA-PASS/pass-docker
# Release Deposit Services 0.0.6-2.2.

- Release notes: https://github.com/OA-PASS/nihms-submission/releases/tag/deposit-services-0.0.6
- Adds passive FTP data ports so pass-docker users can FTP to localhost and retrieve NIH submissions
- Includes a default `packagers.properties` which configures Deposit Services to deposit packages to the local `dspace` and `ftpserver` Docker services
- Includes the ability to override the default `packagers.properties` by bind-mounting a volume, useful for testing staging servers


### Prepare for testing
1. Stop all docker containers, and drop all volumes: `docker-compose down -v`
1. Pull latest images: `docker-compose pull`
1. Start everything up: `docker-compose up -d`
1. Wait for dust to settle: `docker logs -f deposit` will tell you when deposit services and Fedora are ready.
    * Should see `Starting DepositServices (version: 0.0.6-2.2 branch: 151e53205ad5bfb9e7f6c41e2df44a477ad3e5f1 commit: 151e5320 commit date: 2018-06-24T20:17:50-0400)` on startup
    * Should see `INFO  [      ListenerRunner] - Fedora repository is up at 'http://fcrepo:8080/fcrepo/rest/'` when everything is ready.

### Perform a Submission
1. Login to PASS as `nih-user`:`moo`
1. Navigate to the Grants tab, and start a submission for a `National Institutes of Health` grant (e.g. award number `R01EY012124`)
1. I use DOI `10.1039/c7fo01251a`
1. Accept form defaults.  Be sure to accept the JScholarship deposit agreement
1. Attach a bunch of files
1. At this point, clear your scrollback buffer and/or insure you have invoked `docker logs -f deposit` in a terminal window
1. Click submit.

### Expected output
1. You will initially see lots of messages like the following.  These are OK.
    * Precondition for applying the critical path on resource http://fcrepo:8080/fcrepo/rest/submissions/6a/34/f9/55/6a34f955-b72e-4341-87ee-fbe16a564d9b failed.
    * Submission http://fcrepo:8080/fcrepo/rest/submissions/6a/34/f9/55/6a34f955-b72e-4341-87ee-fbe16a564d9b will not be accepted for processing: submitted = false, expected submitted = true
1. You will see Deposit Services accept and process the submission
    * `>>>> Processing Submission http://fcrepo:8080/fcrepo/rest/submissions/6a/34/f9/55/6a34f955-b72e-4341-87ee-fbe16a564d9b`
1. You should see *two* lines similar to the following (grep for `Submitting task ` in your terminal scrollback):
    * `>>>> Submitting task (DepositTask@1bf8e778) for tuple [http://fcrepo:8080/fcrepo/rest/submissions/6a/34/f9/55/6a34f955-b72e-4341-87ee-fbe16a564d9b, http://fcrepo:8080/fcrepo/rest/repositories/41/96/0a/92/41960a92-d3f8-4616-86a6-9e9cadc1a269, http://fcrepo:8080/fcrepo/rest/deposits/2f/2c/78/b5/2f2c78b5-40fb-4563-adb5-d2298e70732a]`
    * Open each `Deposit` resource link in your browser, munging the host name in the url as needed (`http://fcrepo:8080/fcrepo/rest/deposits/2f/2c/78/b5/2f2c78b5-40fb-4563-adb5-d2298e70732a` in this example)

### Verifying output

1. Navigate to one of the `Deposit` resources in your browser
1. Navigate to the `Submission` for the `Deposit`, its aggregatedDepositStatus should be in-progress
    * The files you uploaded should be child resources of the Submission, feel free to download them if you'd like

There are two `Deposit` resources to verify using your browser.  For each `Deposit` resource:

1. Navigate back to the `Deposit`
1. Determine the repository for the deposit by navigating the repository link, then back to the Deposit
1. For the Pub Med repository:
    * the `Deposit` should have status = `submitted`
    * there should be no RepositoryCopy linked to the Deposit
1. For the JScholarship repository:
    * the `Deposit` should have status = `accepted`
    * the `Deposit` should have a value for `depositStatusRef` like `http://pass:8181/swordv2/statement/af2b21ad-3dca-45aa-957f-11b18b5e0ffc.atom`
    * Navigate to the RepositoryCopy
        * Its copy status should be `complete`
        * It should have an externalId like `http://pass:8181/xmlui/handle/123456789/3`
        * It should have an accessUrl like `http://pass:8181/xmlui/handle/123456789/3`
    * Navigate to the access URL, you will have to replace `http` with `https` (this should not be a problem in production, I don't think)
        * View the full record of the DSpace Item.
        * Author information should look sane
        * Other metadata like a DOI should be present

### Bonus verification

* Use your browser to navigate to `ftp://localhost`, and drill down through the directories.  The final directory should be dated with the current date.
    * authenticate as `nihmsftpuser`, `nihmsftppass`
* Drill down into the last directory, and you should see a file named something like: 
    * `nihms-native-2017-07_2018-06-25_00-06-55_6a34f955-b72e-4341-87ee-fbe16a564d9b.tar.gz`
        * where `nihms-native-2017-07` is the packaging spec
        * `2018-06-25_00-06-55` is the date and time (UTC) the package was created
        * `6a34f955-b72e-4341-87ee-fbe16a564d9b` is the final portion of the `Submission` resource's URL.
* Download the file and open it
* You should see all the files you uploaded
* Open `bulk_meta.xml` and you should see 
    * two ISSN's each with their appropriate publication type in individual `<issn>` elements
    * the NLM title abbreviation in the `<journal-id>` element
* Open `manifest.txt` and you should see each file listed
    * exactly one file should be typed as a manuscript
    * the remaining files typed as supplements with unique labels (e.g. `supplement-1`, `supplement-2`.
    * If you provided descriptions or different file types other than the defaults when performing the submission, you should see those values instead.
